extract_region_feature.py

```diff
@@ -0,0 +1,170 @@
+        default="fc7",
+        choices=["fc6", "fc7"],
+    )
+    parser.add_argument(
+        "opts",
+        help="Modify config options using the command-line",
+        default=None,
+        nargs=argparse.REMAINDER,
+    )
+    return parser
+
+
+def extract_feature_on_dataset(model, data_loader, dump_folder, feature_name):
+    for idx, inputs in enumerate(tqdm.tqdm(data_loader)):
+        with torch.no_grad():
+            image_id = inputs[0]["image_id"]
+
+            # compute features and proposals
+            images = model.preprocess_image(inputs)
+            features = model.backbone(images.tensor)
+            proposals, _ = model.proposal_generator(images, features)
+
+            # pooled features and box predictions
+            _, pooled_features, pooled_features_fc6 = model.roi_heads.get_roi_features(
+                features, proposals
+            )
+            predictions = model.roi_heads.box_predictor(pooled_features)
+            cls_probs = F.softmax(predictions[0], dim=-1)
+            cls_probs = cls_probs[:, :-1]  # background is last
+            predictions, r_indices = model.roi_heads.box_predictor.inference(
+                predictions, proposals
+            )
+            # Create Boxes objects from proposals. Since features are extrracted from
+            # the proposal boxes we use them instead of predicted boxes.
+            box_type = type(proposals[0].proposal_boxes)
+            proposal_bboxes = box_type.cat([p.proposal_boxes for p in proposals])
+            proposal_bboxes.tensor = proposal_bboxes.tensor[r_indices]
+            predictions[0].set("proposal_boxes", proposal_bboxes)
+            predictions[0].remove("pred_boxes")
+
+            # postprocess
+            height = inputs[0].get("height")
+            width = inputs[0].get("width")
+            r = postprocessing.detector_postprocess(predictions[0], height, width)
+
+            bboxes = r.get("proposal_boxes").tensor
+            classes = r.get("pred_classes")
+            cls_probs = cls_probs[r_indices]
+            if feature_name == "fc6" and pooled_features_fc6 is not None:
+                pooled_features = pooled_features_fc6[r_indices]
+            else:
+                pooled_features = pooled_features[r_indices]
+
+            assert (
+                bboxes.size(0)
+                == classes.size(0)
+                == cls_probs.size(0)
+                == pooled_features.size(0)
+            )
+
+            # save info and features
+            info = {
+                "bbox": bboxes.cpu().numpy(),
+                "num_boxes": bboxes.size(0),
+                "objects": classes.cpu().numpy(),
+                "image_height": r.image_size[0],
+                "image_width": r.image_size[1],
+                "cls_prob": cls_probs.cpu().numpy(),
+                "features": pooled_features.cpu().numpy()
+            }
+            np.save(os.path.join(dump_folder, str(image_id) + ".npy"), info)
```

Comment:
other parts lgtm, ideally we should test if given the same pre-trained model, using the previous code in mb-benchmark and the current code in D2 should generate the same results. :)